### PR TITLE
[WIP] Add Wasserstein GAN and debug memory leak

### PR DESCRIPTION
**Pull request by rflamary**
This PR introduce the following changes

- [x] New example for Wasserstein GAN
- [x] Correct memory leak in ot.emd2 for torch backend
- [x] Speedup color Domain Adaptation examples for a quicker compilation of the doc

 The memory leak was due to the creation of a new class for each call to ot.emd2 that used torch tensors in their scope which meant that tose tensors could not be de-allocated by the garbage collector. It has been solved by using a torch function that takes both inputs output value and gradients as input bu provides gardients only for the inputs (see set_gradients(self, val, inputs, grads) function on torch backend)

### CODE REVIEW

File: examples/backends/plot_wass2_gan_torch.py [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+# Author: Remi Flamary <remi.flamary@polytechnique.edu>
+#
+# License: MIT License
+
+# sphinx_gallery_thumbnail_number = 2
+
+
+from torch import nn
+import numpy as np
+import matplotlib.pylab as pl
```
Comment (agramfort) [inline examples/backends/plot_wass2_gan_torch.py:54]:
```suggestion
import matplotlib.pyplot as plt
```
this is way more standard

File: examples/backends/plot_wass2_gan_torch.py [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+import ot
+import torch
+
+
+##############################################################################
+# Data generation
+# -------------
```
Comment (agramfort) [inline examples/backends/plot_wass2_gan_torch.py:61]:
```suggestion
# ---------------
```

File: examples/backends/plot_wass2_gan_torch.py [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+
+#%% Data generation
```
Comment (agramfort) [inline examples/backends/plot_wass2_gan_torch.py:63]:
```suggestion
```

File: examples/backends/plot_wass2_gan_torch.py [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+torch.manual_seed(1)
+ncirc = 6
+sig = 0.1
+n = 100
+d = 2
+p = 2
```
Comment (agramfort) [inline examples/backends/plot_wass2_gan_torch.py:69]: can we avoid one letter variables?

n_samples
n_dims
n_features

?

File: examples/backends/plot_wass2_gan_torch.py [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+
+
+##############################################################################
+# Data generation
+# -------------
+
+#%% Data generation
+torch.manual_seed(1)
+ncirc = 6
+sig = 0.1
```
Comment (agramfort) [inline examples/backends/plot_wass2_gan_torch.py:66]: can you can this sigma or std? sig was not obvious to me?

File: examples/backends/plot_wass2_gan_torch.py [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+        output = self.relu(output)
+        output = self.fc3(output)
+        return output
+
+##############################################################################
+# Training the model
+# ------------------
+
+
+#%%
```
Comment (agramfort) [inline examples/backends/plot_wass2_gan_torch.py:120]: if you want to use such vscode / spyder cells you can use them instead of the #### line above.
Sphinx gallery understand them.

Yet I think it should be 
```suggestion
# %%
```

File: examples/backends/plot_wass2_gan_torch.py [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+G = Generator()
+optimizer = torch.optim.RMSprop(G.parameters(), lr=0.001)
+
+# number of iteration and size of the batches
+niter = 500
```
Comment (agramfort) [inline examples/backends/plot_wass2_gan_torch.py:125]:
```suggestion
n_iter = 500
```

File: examples/backends/plot_wass2_gan_torch.py [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+# License: MIT License
+
+# sphinx_gallery_thumbnail_number = 3
+
+
+from torch import nn
+import numpy as np
+import matplotlib.pyplot as pl
+import ot
+import torch
```
Comment (agramfort) [inline examples/backends/plot_wass2_gan_torch.py:56]:
```suggestion
import numpy as np
import matplotlib.pyplot as pl
import torch
from torch import nn
import ot
```

import from more standard to local.

also I note that you don't like my plt :)

File: examples/backends/plot_wass2_gan_torch.py [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+
+# %% Data generation
+# ---------------
```
Comment (agramfort) [inline examples/backends/plot_wass2_gan_torch.py:59]:
```suggestion
# %%
# Data generation
# ----------------
```

File: examples/backends/plot_wass2_gan_torch.py [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+
+torch.manual_seed(1)
+sigma = 0.1
+n_dims = 2
+n_features = 2
+
+
+def get_data(n):
```
Comment (agramfort) [inline examples/backends/plot_wass2_gan_torch.py:67]: n -> n_samples?

File: examples/backends/plot_wass2_gan_torch.py [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+    angle = c * 2 * np.pi
+    x = torch.cat((torch.cos(angle), torch.sin(angle)), 1)
+    x += torch.randn(n, 2) * sigma
+    return x
+
+
+##############################################################################
+# Plot data
+# ---------
+# %% plot the distributions
```
Comment (agramfort) [inline examples/backends/plot_wass2_gan_torch.py:78]:
```suggestion
# %%
# Plot data
# ---------
#
```

File: examples/backends/plot_wass2_gan_torch.py [hunk 12]
```diff
@@ -0,0 +1,194 @@
...
+pl.figure(1)
+pl.scatter(x[:, 0], x[:, 1], label='Data samples from $\mu_d$', alpha=0.5)
+pl.title('Data distribution')
+pl.legend()
+
+
+##############################################################################
+# Generator Model
+# ---------------
+# %% define the MLP model
```
Comment (agramfort) [inline examples/backends/plot_wass2_gan_torch.py:89]: idem

File: examples/backends/plot_wass2_gan_torch.py (hunk 11 above)
Comment (rflamary) [reply examples/backends/plot_wass2_gan_torch.py:78]: github is bugging  I alreday commited some changes but nothing is appearing here. will take into account the other comments and push again

And no I don't like plt, i'm opinionated on this one ;)

**Comment by rflamary** [comment]
https://635-71472695-gh.circle-artifacts.com/0/dev/auto_examples/backends/plot_wass2_gan_torch.html#sphx-glr-auto-examples-backends-plot-wass2-gan-torch-py

Simple 2D Wasserstein GAN in 1 min and 30 seconds.
